version-2/Readme.markdown

```diff
@@ -0,0 +1,406 @@
+# Datagrid Version 2
+
+Datagrid v1 was released on Sep 19 2013 - more than 10 years ago.
+A lot of changes in best practices and available technology
+had happened during this period.
+It caused the library to be designed without support of those technologies
+or their implementation to be suboptimal because of backward compatibility.
+Version 2 addresses all that evolution.
+
+List of things introduces:
+
+1. Use `form_with` instead of `form_for`.
+1. Deprecated `datagrid_order_for`
+1. Ruby endless ranges for range filters.
+1. Modern modular CSS classes.
+1. HTML5 input types: number, date, datetime-local.
+1. Use Hash instead of Array for multiparameter attirubtes.
+1. Native Rails Engines.
+   * while supported, the library was not initially designed for it.
+1. HTML5 data attributes
+1. Inherit `Datagrid::Base` instead of `include Datagrid`
+1. `ApplicationGrid` is recommended base class instead of `BaseGrid`
+1. Remove SASS dependency
+1. Replace `rake datagrid:copy_partials` with `rails g datagrid:views`
+
+## Use form\_with
+
+Rails [deprecates form\_for in favor of form\_with](https://guides.rubyonrails.org/form_helpers.html#using-form-tag-and-form-for).
+
+`datagrid_form_for` is now depreacted if favor of `datagrid_form_with`.
+However, `datagrid_form_for` would also use Rails `form_with` because they share the same view partial.
+
+TODO: update the wiki
+
+``` ruby
+# V1
+datagrid_form_for(@users_grid, url: users_path)
+# V2
+datagrid_form_with(model: @users_grid, url: users_path)
+```
+
+Built-in partial uses `form_with` no matter
+
+[Grep all deprecations](./deprecations.sh).
+
+## Deprecated datagrid\_order\_for
+
+`datagrid_order_for` helper serves no purpose and should not be used directly.
+The recommended way is to include your ordering code directly into `datagrid/head` partial.
+See default [head partial](../app/views/datagrid/_head.html.erb) for example.
+[Grep all deprecations](./deprecations.sh).
+
+## Inherit Datagrid::Base
+
+`include Datagrid` causes method name space to be clamsy.
+Version 2 introduces a difference between the class
+that needs to be inherited and high level namespace (just like most gems do):
+
+``` ruby
+class ApplicationGrid < Datagrid::Base
+end
+```
+
+[Grep all deprecations](./deprecations.sh).
+
+## Endless ranges for range filters
+
+Ruby supports endless ranges now,
+so there is no need to present endless ranges as Hash or Array.
+But it introduces a breaking changes to range filters in Datagrid:
+
+``` ruby
+class UsersGrid < Datagrid::Base
+  filter(:id, :integer, range: true) do |value, scope|
+    # V1 value is  [1, nil]
+    # V2 value is 1..nil
+    scope.where(id: value)
+  end
+end
+
+grid = UsersGrid.new
+grid.id = [1, nil]
+grid.id # V1: [1, nil]
+        # V2: (1..nil)
+```
+
+Version 2 makes an effort to make the transition as smooth as possible to you:
+
+* Old Array format will be converted to new Range format
+* Serialization/Deserialization of Range is help correctly
```

Comment:
What do you mean by "is help correctly"?

Reply:
Typo help => held

Reply:
Please fix the typo.